File: datasets/C++/Chatgpt/Algorithm/medium3.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

int commonChild(string s1, string s2) {
    int n = s1.length();
    vector<vector<int>> dp(n + 1, vector<int>(n + 1, 0));

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            if (s1[i] == s2[j])
                dp[i + 1][j + 1] = dp[i][j] + 1;
            else
                dp[i + 1][j + 1] = max(dp[i][j + 1], dp[i + 1][j]);
        }
    }
    return dp[n][n];
}

int main() {
    string s1, s2;
    cout << "Enter first string: ";
    cin >> s1;
    cout << "Enter second string: ";
    cin >> s2;

    cout << "Length of common child: " << commonChild(s1, s2) << endl;
    return 0;
}

File: datasets/Python/Chatgpt/Algorithm/medium3.py
def commonChild(s1, s2):
    n = len(s1)
    dp = [[0] * (n + 1) for _ in range(n + 1)]

    for i in range(n):
        for j in range(n):
            if s1[i] == s2[j]:
                dp[i+1][j+1] = dp[i][j] + 1
            else:
                dp[i+1][j+1] = max(dp[i][j+1], dp[i+1][j])

    return dp[n][n]

# Manual input
s1 = input("Enter first string: ").strip()
s2 = input("Enter second string: ").strip()

print("Length of common child:", commonChild(s1, s2))
